needle-notion-connector/src/server/backend/connectors-backend.ts

```diff
@@ -0,0 +1,154 @@
+import { Client as NotionClient } from "@notionhq/client";
+import type {
+  DatabaseObjectResponse,
+  PageObjectResponse,
+} from "@notionhq/client/build/src/api-endpoints";
+
+export async function createNotionConnector(
+  request: CreateConnectorRequest,
+  session: Session,
+) {
+  const connector = await createConnector(
+    {
+      name: request.name,
+      cronJob: request.cronJob,
+      cronJobTimezone: request.cronJobTimezone,
+      collectionIds: [request.collectionId],
+      credentials: request.notionToken.access_token,
+    },
+    session.id,
+  );
+
+  const pagesToInsert = request.notionPages.map((p) => ({
+    ndlConnectorId: connector.id,
+    ndlFileId: createNeedleFileId(),
+    notionUrl: p.url,
+    notionPageId: p.id,
+    notionLastEditedTime: p.last_edited_time,
+  }));
+  await db.insert(notionPagesTable).values(pagesToInsert);
+
+  const connectorsToInsert = request.notionPages.map((p) => ({
+    ndlConnectorId: connector.id,
+    notionWorkspaceId: request.notionToken.workspace_id,
+    notionWorkspaceName: request.notionToken.workspace_name,
+    notionUserId: request.notionToken.owner.user.id,
+    notionPageId: p.id,
+  }));
+  await db.insert(notionConnectorsTable).values(connectorsToInsert);
+
+  await runNotionConnector({ connectorId: connector.id }, session);
+
+  return connector;
+}
+
+export async function listNotionConnectors(session: Session) {
+  return await listConnectors(session.id);
+}
+
+export async function getNotionConnector(
+  { connectorId }: ConnectorRequest,
+  session: Session,
+) {
+  const connector = await getConnector(connectorId, session.id);
+
+  const files = await db
+    .select()
+    .from(notionPagesTable)
+    .where(eq(notionPagesTable.ndlConnectorId, connectorId));
+
+  return { ...connector, files };
+}
+
+export async function deleteNotionConnector(
+  { connectorId }: ConnectorRequest,
+  session: Session,
+) {
+  const connector = await deleteConnector(connectorId, session.id);
+  await db
+    .delete(notionPagesTable)
+    .where(eq(notionPagesTable.ndlConnectorId, connectorId));
+  await db
+    .delete(notionConnectorsTable)
+    .where(eq(notionConnectorsTable.ndlConnectorId, connectorId));
+  return connector;
+}
+
+export async function runNotionConnector(
+  { connectorId }: ConnectorRequest,
+  session?: Session,
+) {
+  let connector;
+  if (session) {
+    connector = await getConnector(connectorId, session.id);
+  }
+  if (!connector?.credentials) {
+    throw new Error("notion access token is not found");
+  }
+
+  const notion = new NotionClient({ auth: connector.credentials });
+  const notionSearchResponse = await notion.search({});
+  const notionResults = notionSearchResponse.results as (
+    | PageObjectResponse
+    | DatabaseObjectResponse
+  )[];
+
+  const ndlPages = await db
+    .select()
+    .from(notionPagesTable)
+    .where(eq(notionPagesTable.ndlConnectorId, connectorId));
+
+  const pagesToCreate = notionResults.filter((r) => {
```

Comment:
This seems like a nested loop with O(n^2) complexity. Alternative would be to create a map and compute diff based on them, which would be O(n) complexity. I like to follow this pseudo-code for that. I think set based diff reads easier. However up to you to - I'm ok with going with this solution.


```js
// currently tracked data
cur_pages = db.list_pages
cur_pages_map = new Map(cur_pages.map(p => (p.id, p)))
cur_page_ids = cur_pages_map.keys()

// live data
live_pages = notion.search()
live_pages_map = new Map(live_pages.map(p => (p.id, p)))
live_page_ids = live_pages_map.keys()

// diff
page_ids_to_create = live_page_ids - cur_page_ids 			// set difference
page_ids_to_delete = cur_page_ids - live_page_ids 			// set difference
page_ids_to_maybe_update = cur_page_ids & live_page_ids 	// set intersection

// execute create actions

// execute delete actions

// iterate over maybe_update actions and update if data changed

```